### PR TITLE
Fix for issue #225

### PR DESCRIPTION
**Pull request by jpthompson17**
**Issue Number.** Is this pull request related to any outstanding issues? If so, list the issue number.  
#225 
 
**Describe the changes made.** A clear and concise description of what the problem is and what you did to fix it. E.g. [...] was happening and I've changed [...] to fix it.  

This PR addresses problems initially reported by @jeremyleung521 in https://github.com/westpa/westpa/pull/217#issuecomment-1018735223. It also includes a small performance enhancement of `Walker.children`.

**Goals and Outstanding Issues.** A clear and concise list of goals (to be) accomplished.
- Handle the case that the final iteration in the west.h5 file (`west_current_iteration`) is complete (e.g. after `w_truncate`ing the run).
- Avoid an `IndexError` when the parent of a walker is an initial state that was *not* generated from a basis state (e.g., `ISTATE_TYPE_START`).

**Major files changed.**  
- westpa/analysis/core.py

**Status.**
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

**Additional context.** Add any other context or screenshots about the pull request here.

### CODE REVIEW

**Comment by jeremyleung521** [comment]
I'm having trouble with [these same test files](https://github.com/westpa/westpa2_tutorials/tree/main/basic_nacl/w_crawl) again. I tried to read iteration 2, walker 18 with the HDF5MDTrajectory(), which has the parent iteration 1, walker 7. The pcoords, as far as I can see, have the same dimensions. I can read iteration 2 fine though! I think this is a bug from @SHZ66's HDF5MDTrajectory object. 

This is independent of whether it's from a bstate or an sstate.  (Try iteration 2, walker 30 if you want something from the bstates)

Screenshot 2022-01-26 at 2 57 51 PM

File: src/westpa/analysis/core.py [hunk 1]
```diff
@@ -11,12 +11,12 @@
...
 
 class Run:
-    """A completed WESTPA simulation run.
+    """A read-only view of a WESTPA simulation run.
 
     Parameters
     ----------
     h5filename : str or file-like object, default 'west.h5'
-        Pathname or stream of a WESTPA HDF5 file.
+        Pathname or stream of a WEST HDF5 file.
```
Comment (jeremyleung521) [inline src/westpa/analysis/core.py:19]: Just to be in line with the terminology of the WESTPA 2.0 paper. 

[...] of a main WESTPA HDF5 data file.

File: src/westpa/analysis/core.py [hunk 2]
```diff
@@ -38,13 +38,13 @@ def open(cls, h5filename='west.h5'):
         Parameters
         ----------
         h5filename : str or file-like object, default 'west.h5'
-            Pathname or stream of a WESTPA HDF5 file.
+            Pathname or stream of a WEST HDF5 file.
```
Comment (jeremyleung521) [inline src/westpa/analysis/core.py:41]: Just to be in line with the terminology of the WESTPA 2.0 paper.

[...] of a main WESTPA HDF5 data file.

File: src/westpa/analysis/core.py [hunk 3]
```diff
@@ -38,13 +38,13 @@ def open(cls, h5filename='west.h5'):
...
 
         """
         return cls(h5filename)
 
     def close(self):
-        """Close the Run instance by closing the underlying WESTPA HDF5 file."""
+        """Close the Run instance by closing the underlying WEST HDF5 file."""
```
Comment (jeremyleung521) [inline src/westpa/analysis/core.py:47]: Just to be in line with the terminology of the WESTPA 2.0 paper.

[...] of a main WESTPA HDF5 data file.

File: src/westpa/analysis/core.py (hunk 1 above)
Comment (jpthompson17) [reply src/westpa/analysis/core.py:19]: Oops, I was trying to be in line with https://westpa.github.io/westpa/users_guide/hdf5.html. I'll change this to be consistent with the paper.

**Comment by jpthompson17** [comment]
> I'm having trouble with [these same test files](https://github.com/westpa/westpa2_tutorials/tree/main/basic_nacl/w_crawl) again. I tried to read iteration 2, walker 18 with the HDF5MDTrajectory(), which has the parent iteration 1, walker 7. The pcoords, as far as I can see, have the same dimensions. I think this is a bug from @SHZ66's HDF5MDTrajectory object.
> 
> This is independent of whether it's from a bstate or an sstate. (Try iteration 2, walker 30 if you want something from the bstates)
> 
> Screenshot 2022-01-26 at 2 57 51 PM

I'm looking into this. Part of the issue seems to be that `WESTIterationFile.read_as_traj()` returns a `WESTTrajectory` rather than an ordinary `mdtraj.Trajectory`. (I missed the type issue because the `repr()` of a `WESTTrajectory` makes it look like an ordinary MDTraj trajectory, thanks partly to the hard-coded class name at https://github.com/mdtraj/mdtraj/blob/51779e2c820df155782ec24f6d9e2701939cc539/mdtraj/core/trajectory.py#L948.) The `WESTTrajectory` class overrides the `join()` method of `mdtraj.Trajectory`, which is where the exception is raised.

I'm thinking we may need to (essentially) upcast the `WESTTrajectory` objects in the HDF5 trajectory reader (i.e., `traj` and `frame`) to MDTraj trajectories.

File: src/westpa/analysis/core.py (hunk 1 above)
Comment (jeremyleung521) [reply src/westpa/analysis/core.py:19]: > Oops, I was trying to be in line with https://westpa.github.io/westpa/users_guide/hdf5.html. I'll change this to be consistent with the paper.

No worries. That was written when WESTPA was still called 'west' and had no concept of per-iteration h5s, so it's a bit outdated.

**Comment by jeremyleung521** [comment]
> > I'm having trouble with [these same test files](https://github.com/westpa/westpa2_tutorials/tree/main/basic_nacl/w_crawl) again. I tried to read iteration 2, walker 18 with the HDF5MDTrajectory(), which has the parent iteration 1, walker 7. The pcoords, as far as I can see, have the same dimensions. I think this is a bug from @SHZ66's HDF5MDTrajectory object.
> > This is independent of whether it's from a bstate or an sstate. (Try iteration 2, walker 30 if you want something from the bstates)
> > Screenshot 2022-01-26 at 2 57 51 PM
> 
> I'm looking into this. Part of the issue seems to be that `WESTIterationFile.read_as_traj()` returns a `WESTTrajectory` rather than an ordinary `mdtraj.Trajectory`. (I missed the type issue because the `repr()` of a `WESTTrajectory` makes it look like an ordinary MDTraj trajectory, thanks partly to the hard-coded class name at https://github.com/mdtraj/mdtraj/blob/51779e2c820df155782ec24f6d9e2701939cc539/mdtraj/core/trajectory.py#L948.) The `WESTTrajectory` class overrides the `join()` method of `mdtraj.Trajectory`, which is where the exception is raised.
> 
> I'm thinking we may need to (essentially) upcast the `WESTTrajectory` objects in the HDF5 trajectory reader (i.e., `traj` and `frame`) to MDTraj trajectories.

Seems like it's fixed with #227. Perfectly fine with merging this PR after those docstrings are fixed! 

For the record, I tesed with a branch with both PRs and they don't seem to interfere with each other.

**Comment by jpthompson17** [comment]
@jeremyleung521 I revised the docstrings for `Run.__init__()` and `Run.open()`, and changed "WEST" back to "WESTPA" in the docstring for `Run.close()`.

**Comment by jeremyleung521** [comment]
> @jeremyleung521 I revised the docstrings for `Run.__init__()` and `Run.open()`, and changed "WEST" back to "WESTPA" in the docstring for `Run.close()`.

Sounds good. I will merge this and #227 in once the CI's complete!

**Pull request merged**